MessageSigning/MessageSigning.md

```diff
@@ -0,0 +1,437 @@
+
+You can find more complete definitions of `protected` and `unprotected` in [RFC 8152 section-3](https://tools.ietf.org/html/rfc8152#section-3) and you can see some the reserved entries (called `Generic_Headers`) in the maps in [RFC 8152 section-3.1](https://tools.ietf.org/html/rfc8152#section-3.1).
+
+Note: `payload` can be `nil`. This means that the payload is known by both the signer and the verifier and therefore doesn't need to be encoded.
+
+For your convenience, the structure is provided here:
+
+```
+empty_or_serialized_map = bstr .cbor header_map / bstr .size 0
+
+header_map = {
+    Generic_Headers,   ; reserved headers (see COSE section 3.1)
+    * label => values  ; any number of int/string labels for application-specific purpose.
+}
+
+Headers = (
+    protected : empty_or_serialized_map,
+    unprotected : header_map
+)
+
+; signature layer
+COSE_Signature =  [
+    Headers,
+    signature : bstr
+]
+
+; if signing with just ONE key
+COSE_Sign1 = [
+    Headers,
+    payload : bstr / nil,
+    signature : bstr
+]
+
+# if signing with >1 key
+COSE_Sign = [
+    Headers,
+    payload : bstr / nil,
+    signatures : [+ COSE_Signature]
+]
+
+signed_message = COSE_SIGN / COSE_Sign1
+```
+
+#### Signing and Verification target format
+
+Instead of signing the full structure, we instead sign the following type which is derived from the structure
+
+```
+Sig_structure = [
+  context : "Signature" / "Signature1" / "CounterSignature",    ; explained below
+  body_protected : empty_or_serialized_map,                     ; protected from layer 1
+  ? sign_protected : empty_or_serialized_map,                   ; not present in Sign1 case
+  external_aad : bstr,                                          ; explanation below
+  payload : bstr
+]
+```
+
+The `external_aad` allows an application to ask the user to sign some extra data but NOT put it inside the COSE structure (only as part of the data to sign). Defaults to `h''`. You can read more about this at [RFC 8152 section-4.3](https://tools.ietf.org/html/rfc8152#section-4.3).
+
+The `context` is meant to encode what structure was used for the COSE request. `CounterSignature` is explained in a later section of this specification.
+
+#### Signing and Verification process
+
+To be able to effectively verify we need two things:
+
+1) P1 - (optional) knowledge of the relation of a public key and a Cardano address
+1) P2 - Knowledge of which algorithm was used to sign
+
+For `P1`, just knowing the public key is not sufficient to know the address (which we need to know if we want to prove ownership of a UTXO or transaction). To resolve this, first we (optionally) add to the header an extra field.
+
+```
+? address: bstr,
+? address_pubkey: bstr,
```

Comment:
For a different application, I am interested in the case where A provides B with a signature of a piece of data, but B does not (yet) know A's public key, so cannot actually perform verification. So we want to "annotate" the signature with the key that provided it (under some circumstances this might be an information leak, but not for this application).

However, in this case the public key need not correspond to an address. So my suggestion would simply be to have this header be `signing_pubkey` instead, which hopefully doesn't affect the clarity in the case where it also corresponds to an address, but would make more sense for non-address-related usecases.

Reply:
Moreover, reading through some of the COSE spec, it *seems* to me that there are already appropriate places for putting this information. For example, the `kid` unprotected header:
```
   kid:  This parameter identifies one piece of data that can be used as
      input to find the needed cryptographic key.  The value of this
      parameter can be matched against the 'kid' member in a COSE_Key
      structure.  Other methods of key distribution can define an
      equivalent field to be matched.  Applications MUST NOT assume that
      'kid' values are unique.  There may be more than one key with the
      same 'kid' value, so all of the keys associated with this 'kid'
      may need to be checked.  The internal structure of 'kid' values is
      not defined and cannot be relied on by applications.  Key
      identifier values are hints about which key to use.  This is not a
      security critical field.  For this reason, it can be placed in the
      unprotected headers bucket.
```
In particular, it seems like it would be conformant to put the signing key itself in the `kid` header.